baseapp/baseapp.go

```diff
@@ -654,12 +700,17 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 		anteEvents = events.ToABCIEvents()
 	}
 
-	// TODO remove nil check when implemented
-	if mode == runTxModeCheck && app.mempool != nil {
-		err = app.mempool.Insert(ctx, tx.(mempool.Tx))
+	if mode == runTxModeCheck {
+		err = app.mempool.Insert(ctx, tx)
 		if err != nil {
 			return gInfo, nil, anteEvents, priority, err
 		}
+	} else if mode == runTxModeDeliver {
+		err = app.mempool.Remove(tx)
```

Comment:
you remove the tx in the process proposal phase so this wold always error? I think removing here is better than in processproposal, or is there a design im missing?

Reply:
Only invalid txs are removed in the default ProcessProposal implementation.  Also we ignore `mempool.ErrTxNotFound` errors.

Reply:
the invalid txs are defined by the antehandler returning an error on them